File: ai-ml/tests/phase_6_2/test_api_endpoints.py
#!/usr/bin/env python3
"""
API Endpoint Testing Script
Test the REGIQ AI/ML API endpoints
"""

import requests
import json
import time

# API base URL
BASE_URL = "http://localhost:8000"
API_PREFIX = "/api/v1"

def test_health_endpoint():
    """Test the health endpoint"""
    print("Testing health endpoint...")
    try:
        response = requests.get(f"{BASE_URL}/health")
        if response.status_code == 200:
            print("✅ Health endpoint: OK")
            print(f"   Response: {response.json()}")
            return True
        else:
            print(f"❌ Health endpoint failed with status {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Health endpoint error: {e}")
        return False

def test_root_endpoint():
    """Test the root endpoint"""
    print("\nTesting root endpoint...")
    try:
        response = requests.get(f"{BASE_URL}/")
        if response.status_code == 200:
            print("✅ Root endpoint: OK")
            print(f"   Response: {response.json()}")
            return True
        else:
            print(f"❌ Root endpoint failed with status {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Root endpoint error: {e}")
        return False

def test_docs_endpoint():
    """Test the docs endpoint"""
    print("\nTesting docs endpoint...")
    try:
        response = requests.get(f"{BASE_URL}/docs")
        if response.status_code == 200:
            print("✅ Docs endpoint: OK")
            return True
        else:
            print(f"❌ Docs endpoint failed with status {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Docs endpoint error: {e}")
        return False

def test_regulatory_intelligence_endpoints():
    """Test regulatory intelligence endpoints (structure only)"""
    print("\nTesting regulatory intelligence endpoint structure...")
    
    endpoints = [
        ("/regulatory-intelligence/documents/analyze", "POST"),
        ("/regulatory-intelligence/summarize", "POST"),
        ("/regulatory-intelligence/qa", "POST"),
        ("/regulatory-intelligence/search", "POST")
    ]
    
    success_count = 0
    for endpoint, method in endpoints:
        try:
            url = f"{BASE_URL}{API_PREFIX}{endpoint}"
            if method == "POST":
                # Test with empty payload to check if endpoint exists
                response = requests.post(url, json={}, timeout=5)
            else:
                response = requests.get(url, timeout=5)
            
            # We expect 422 (validation error) or 401 (auth error) if endpoint exists
            # What we don't want is 404 (endpoint not found)
            if response.status_code != 404:
                print(f"✅ {method} {endpoint}: Endpoint exists (status {response.status_code})")
                success_count += 1
            else:
                print(f"❌ {method} {endpoint}: Endpoint not found")
        except requests.exceptions.ConnectionError:
            print(f"❌ {method} {endpoint}: Connection error")
        except Exception as e:
            print(f"✅ {method} {endpoint}: Endpoint exists (error: {type(e).__name__})")
            success_count += 1
    
    return success_count == len(endpoints)

def test_bias_analysis_endpoints():
    """Test bias analysis endpoints (structure only)"""
    print("\nTesting bias analysis endpoint structure...")
    
    endpoints = [
        ("/bias-analysis/models/upload", "POST"),
        ("/bias-analysis/analyze", "POST"),
        ("/bias-analysis/results/test_id", "GET"),
        ("/bias-analysis/reports/generate", "POST")
    ]
    
    success_count = 0
    for endpoint, method in endpoints:
        try:
            url = f"{BASE_URL}{API_PREFIX}{endpoint}"
            if method == "POST":
                response = requests.post(url, json={}, timeout=5)
            else:
                response = requests.get(url, timeout=5)
            
            # We expect 422 (validation error) or 401 (auth error) if endpoint exists
            if response.status_code != 404:
                print(f"✅ {method} {endpoint}: Endpoint exists (status {response.status_code})")
                success_count += 1
            else:
                print(f"❌ {method} {endpoint}: Endpoint not found")
        except requests.exceptions.ConnectionError:
            print(f"❌ {method} {endpoint}: Connection error")
        except Exception as e:
            print(f"✅ {method} {endpoint}: Endpoint exists (error: {type(e).__name__})")
            success_count += 1
    
    return success_count == len(endpoints)

def test_risk_simulator_endpoints():
    """Test risk simulator endpoints (structure only)"""
    print("\nTesting risk simulator endpoint structure...")
    
    endpoints = [
        ("/risk-simulator/setup", "POST"),
        ("/risk-simulator/run/test_id", "POST"),
        ("/risk-simulator/stream/test_id", "GET"),
        ("/risk-simulator/scenarios", "GET"),
        ("/risk-simulator/scenarios", "POST")
    ]
    
    success_count = 0
    for endpoint, method in endpoints:
        try:
            url = f"{BASE_URL}{API_PREFIX}{endpoint}"
            if method == "POST":
                response = requests.post(url, json={}, timeout=5)
            else:
                response = requests.get(url, timeout=5)
            
            # We expect 422 (validation error) or 401 (auth error) if endpoint exists
            if response.status_code != 404:
                print(f"✅ {method} {endpoint}: Endpoint exists (status {response.status_code})")
                success_count += 1
            else:
                print(f"❌ {method} {endpoint}: Endpoint not found")
        except requests.exceptions.ConnectionError:
            print(f"❌ {method} {endpoint}: Connection error")
        except Exception as e:
            print(f"✅ {method} {endpoint}: Endpoint exists (error: {type(e).__name__})")
            success_count += 1
    
    return success_count == len(endpoints)

def test_report_generation_endpoints():
    """Test report generation endpoints (structure only)"""
    print("\nTesting report generation endpoint structure...")
    
    endpoints = [
        ("/reports/create", "POST"),
        ("/reports/templates", "GET"),
        ("/reports/templates", "POST"),
        ("/reports/export/test_id", "GET"),
        ("/reports/status/test_id", "GET")
    ]
    
    success_count = 0
    for endpoint, method in endpoints:
        try:
            url = f"{BASE_URL}{API_PREFIX}{endpoint}"
            if method == "POST":
                response = requests.post(url, json={}, timeout=5)
            else:
                response = requests.get(url, timeout=5)
            
            # We expect 422 (validation error) or 401 (auth error) if endpoint exists
            if response.status_code != 404:
                print(f"✅ {method} {endpoint}: Endpoint exists (status {response.status_code})")
                success_count += 1
            else:
                print(f"❌ {method} {endpoint}: Endpoint not found")
        except requests.exceptions.ConnectionError:
            print(f"❌ {method} {endpoint}: Connection error")
        except Exception as e:
            print(f"✅ {method} {endpoint}: Endpoint exists (error: {type(e).__name__})")
            success_count += 1
    
    return success_count == len(endpoints)

def main():
    """Main test function"""
    print(f"Testing REGIQ AI/ML API at {BASE_URL}")
    print("=" * 50)
    
    # Test basic endpoints
    health_ok = test_health_endpoint()
    root_ok = test_root_endpoint()
    docs_ok = test_docs_endpoint()
    
    # Test service endpoints
    ri_ok = test_regulatory_intelligence_endpoints()
    ba_ok = test_bias_analysis_endpoints()
    rs_ok = test_risk_simulator_endpoints()
    rg_ok = test_report_generation_endpoints()
    
    print("\n" + "=" * 50)
    print("TEST SUMMARY")
    print("=" * 50)
    print(f"Health Endpoint: {'✅ PASS' if health_ok else '❌ FAIL'}")
    print(f"Root Endpoint: {'✅ PASS' if root_ok else '❌ FAIL'}")
    print(f"Docs Endpoint: {'✅ PASS' if docs_ok else '❌ FAIL'}")
    print(f"Regulatory Intelligence: {'✅ PASS' if ri_ok else '❌ FAIL'}")
    print(f"Bias Analysis: {'✅ PASS' if ba_ok else '❌ FAIL'}")
    print(f"Risk Simulator: {'✅ PASS' if rs_ok else '❌ FAIL'}")
    print(f"Report Generation: {'✅ PASS' if rg_ok else '❌ FAIL'}")
    
    overall_success = all([health_ok, root_ok, docs_ok, ri_ok, ba_ok, rs_ok, rg_ok])
    print("\n" + "=" * 50)
    if overall_success:
        print("🎉 ALL TESTS PASSED! API is running correctly.")
    else:
        print("⚠️  Some tests failed. Please check the output above.")
    print("=" * 50)
    
    return overall_success

if __name__ == "__main__":
    main()

File: backend/test-api-endpoints.js
#!/usr/bin/env node

/**
 * API Endpoint Alignment Test
 * Verifies all frontend ↔ backend endpoints are working
 */

const axios = require('axios');

// Configuration
const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:3000';
const TEST_TIMEOUT = 5000;

// Test cases organized by category
const testCases = [
  {
    category: 'Health Check',
    tests: [
      {
        name: 'Backend Health',
        method: 'GET',
        endpoint: '/health',
        expectedStatus: 200
      }
    ]
  },
  
  {
    category: 'Regulatory Intelligence',
    tests: [
      {
        name: 'Get Regulations',
        method: 'GET',
        endpoint: '/regulatory/regulations',
        expectedStatus: 200
      },
      {
        name: 'Get Regulation Categories',
        method: 'GET',
        endpoint: '/regulatory/regulations/categories',
        expectedStatus: 200
      },
      {
        name: 'Get Regulation Deadlines',
        method: 'GET',
        endpoint: '/regulatory/regulations/deadlines',
        expectedStatus: 200
      }
    ]
  },
  
  {
    category: 'Bias Analysis',
    tests: [
      {
        name: 'Get Bias Reports',
        method: 'GET',
        endpoint: '/api/bias/reports',
        expectedStatus: 200
      },
      {
        name: 'Get Bias Scoring',
        method: 'GET',
        endpoint: '/api/bias/scoring',
        expectedStatus: 200
      },
      {
        name: 'Get Bias Visualization',
        method: 'GET',
        endpoint: '/api/bias/visualization',
        expectedStatus: 200
      }
    ]
  },
  
  {
    category: 'Risk Simulation',
    tests: [
      {
        name: 'Get Risk Simulations',
        method: 'GET',
        endpoint: '/api/risk/simulations',
        expectedStatus: 200
      },
      {
        name: 'Get Risk Scenarios',
        method: 'GET',
        endpoint: '/api/risk/scenarios',
        expectedStatus: 200
      },
      {
        name: 'Get Regulatory Frameworks',
        method: 'GET',
        endpoint: '/api/risk/frameworks',
        expectedStatus: 200
      }
    ]
  },
  
  {
    category: 'Report Generation',
    tests: [
      {
        name: 'Get Reports',
        method: 'GET',
        endpoint: '/api/reports',
        expectedStatus: 200
      },
      {
        name: 'Get Report Glossary',
        method: 'GET',
        endpoint: '/api/reports/glossary',
        expectedStatus: 200
      },
      {
        name: 'Get Report Templates',
        method: 'GET',
        endpoint: '/api/reports/templates',
        expectedStatus: 200
      }
    ]
  },
  
  {
    category: 'User Management',
    tests: [
      {
        name: 'Get Users (requires auth)',
        method: 'GET',
        endpoint: '/api/users',
        expectedStatus: 401, // Requires authentication
        note: 'This is expected - needs valid JWT token'
      },
      {
        name: 'Get User Preferences (requires auth)',
        method: 'GET',
        endpoint: '/api/users/preferences',
        expectedStatus: 401, // Requires authentication
        note: 'This is expected - needs valid JWT token'
      }
    ]
  },
  
  {
    category: 'Notifications',
    tests: [
      {
        name: 'Get Notifications (requires auth)',
        method: 'GET',
        endpoint: '/api/notifications',
        expectedStatus: 401, // Requires authentication
        note: 'This is expected - needs valid JWT token'
      },
      {
        name: 'Get Notification Preferences (requires auth)',
        method: 'GET',
        endpoint: '/api/notifications/preferences',
        expectedStatus: 401, // Requires authentication
        note: 'This is expected - needs valid JWT token'
      }
    ]
  }
];

// Results tracking
const results = {
  total: 0,
  passed: 0,
  failed: 0,
  errors: []
};

/**
 * Test single endpoint
 */
async function testEndpoint(category, test) {
  const { name, method, endpoint, expectedStatus, note } = test;
  const url = `${BACKEND_URL}${endpoint}`;
  
  try {
    const response = await axios({
      method,
      url,
      timeout: TEST_TIMEOUT,
      validateStatus: () => true // Don't throw on error status
    });
    
    results.total++;
    
    if (response.status === expectedStatus) {
      console.log(`✅ ${category}: ${name}`);
      if (note) {
        console.log(`   ℹ️  Note: ${note}`);
      }
      results.passed++;
      return true;
    } else {
      console.log(`⚠️  ${category}: ${name} - Status: ${response.status} (expected ${expectedStatus})`);
      results.failed++;
      results.errors.push({
        category,
        name,
        type: 'status_mismatch',
        expected: expectedStatus,
        actual: response.status
      });
      return false;
    }
  } catch (error) {
    results.total++;
    results.failed++;
    console.log(`❌ ${category}: ${name} - ${error.message}`);
    results.errors.push({
      category,
      name,
      type: 'error',
      message: error.message
    });
    return false;
  }
}

/**
 * Run all tests
 */
async function runTests() {
  console.log('╔════════════════════════════════════════════╗');
  console.log('║     REGIQ API Endpoint Alignment Test     ║');
  console.log('╚════════════════════════════════════════════╝\n');
  console.log(`Backend URL: ${BACKEND_URL}`);
  console.log(`Timeout: ${TEST_TIMEOUT}ms\n`);
  
  // First check if backend is running
  console.log('📋 Checking backend availability...\n');
  try {
    await axios.get(`${BACKEND_URL}/health`, { timeout: 5000 });
    console.log('✅ Backend is running\n');
  } catch (error) {
    console.log('❌ Backend not available!');
    console.log(`   Error: ${error.message}`);
    console.log('\n   Please start the backend server:');
    console.log('   cd backend && npm run dev\n');
    return;
  }
  
  // Run tests by category
  console.log('📋 Testing API endpoints...\n');
  
  for (const category of testCases) {
    console.log(`\n${category.category}:`);
    console.log('─'.repeat(40));
    
    for (const test of category.tests) {
      await testEndpoint(category.category, test);
      await new Promise(resolve => setTimeout(resolve, 200)); // Small delay between tests
    }
  }
  
  // Print summary
  console.log('\n╔════════════════════════════════════════════╗');
  console.log('║              TEST SUMMARY                  ║');
  console.log('╚════════════════════════════════════════════╝\n');
  console.log(`Total Tests: ${results.total}`);
  console.log(`Passed: ${results.passed} ✅`);
  console.log(`Failed: ${results.failed} ❌`);
  console.log(`Success Rate: ${((results.passed / results.total) * 100).toFixed(1)}%\n`);
  
  if (results.failed > 0) {
    console.log('📝 Issues Found:\n');
    results.errors.forEach((error, index) => {
      console.log(`${index + 1}. ${error.category} - ${error.name}`);
      if (error.type === 'status_mismatch') {
        console.log(`   Expected status ${error.expected}, got ${error.actual}`);
      } else {
        console.log(`   Error: ${error.message}`);
      }
    });
    
    console.log('\n💡 Recommendations:');
    console.log('  • Check if backend routes are properly configured');
    console.log('  • Verify database connection if getting 500 errors');
    console.log('  • Check CORS configuration for cross-origin requests');
    console.log('  • Review API_ENDPOINT_MAPPING_COMPLETE.md for details\n');
  } else {
    console.log('🎉 All endpoints are working correctly!\n');
    console.log('✅ Priority 2: API Endpoint Alignment - COMPLETE\n');
  }
}

// Run the tests
runTests().catch(console.error);
